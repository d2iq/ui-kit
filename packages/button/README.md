# Buttons

Buttons are used for action items. Some examples include Create, Save, Delete. Buttons make it obvious to the user that they can take an action.

Do not use buttons as navigational elements. Instead, use anchors because it takes the user to a new page and is not associated with an action.

Each page will likely have one primary button. Any remaining calls-to-action are represented as secondary buttons.

## Button states and types

### Types

**Primary**: Primary call to action button. Should stand out above all other actions. Do not put two primary buttons next to each other. Examples include Save, Deploy.

**Secondary**: Most buttons on a page are likely this style.

**Danger**: Destructive buttons are used to highlight something dangerous that can't be undone e.g. "Delete".

**Icon only**: Use when an icon is enough to represent the action e.g. "Create a Service". Should be accompanied by a tooltip.

**Icon with label**: If there is space and we feel an icon by itself is not obvious enough e.g. "Kill Task"

### States

**Disabled**: Disable a button when the action is not applicable but we want to show the user that it may be possible. Disabled buttons should have a tooltip.

**Processing**: When user clicks or taps a button and an API request is initiated we should provide feedback so they know their action was received.

**Hidden:** Hide a button when there is no way a user can interact with the button.

## Using buttons as links

Sometimes there is a button that is actually a link to somewhere else. If the appropriate props are set, the button is rendered as an anchor tag, but it is styled like a button.

It is not recommended to use a button as a link when it triggers a dropdown. It would be frustrating if a user clicked a button and expected to see a dropdown appear, but they are taken to a new page.

## Button content

Button labels tell users what will happen when they click the button. Use verbs that describe the action. Acceptable labels include:

- New \$object
- Create \$object
- Edit \$object
- Save \$object
- Delete \$object
- Deploy, Scale, Restart

Use sentence-style capitalization (only the first word in a phrase and any proper nouns capitalized) and no more than three words for button labels.

For Sets of Buttons, use specific labels, such as Save or Cancel, instead of using Yes and No. This is particularly helpful when the user is confirming an action.

## ResetButton

A `ResetButton` component is essentially an unstyled HTML `<button>` element. It can be used for an element that is semantically a button, but is not intended to look like a button.

## Feedback

Buttons should have hover, active, focus, disabled states.

When a user interacts with a button, either via click or tap, and an action is being completed, e.g. an API call, the button should change to disabled and provide feedback via the label to the user that it received the interaction and is busy performing the action e.g. “Saving…”

## Tooltips

Buttons may or may not require tooltips to help communicate their intent.

Icon buttons that do not have labels should always have a tooltip so the user knows what will happen when they click it.

Disabled buttons should always have a tooltip to explain why it is disabled so the user is aware.
