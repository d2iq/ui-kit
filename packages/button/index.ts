export { default as PrimaryButton } from "./components/PrimaryButton";
export { default as SecondaryButton } from "./components/SecondaryButton";
export { default as StandardButton } from "./components/StandardButton";
export { default as SuccessButton } from "./components/SuccessButton";
export { default as DangerButton } from "./components/DangerButton";

export { default as PrimaryDropdownButton } from "./components/PrimaryDropdownButton";
export { default as SecondaryDropdownButton } from "./components/SecondaryDropdownButton";
export { default as StandardDropdownButton } from "./components/StandardDropdownButton";
export { default as SuccessDropdownButton } from "./components/SuccessDropdownButton";
export { default as DangerDropdownButton } from "./components/DangerDropdownButton";

export { default as ResetButton } from "./components/ResetButton";
