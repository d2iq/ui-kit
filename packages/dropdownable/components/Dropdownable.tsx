import React from "react";
import styled, { css } from "react-emotion";

import Overlay from "../../shared/components/Overlay";
import DropdownContents from "./DropdownContents";
import resizeEventManager from "../../utilities/resizeEventManager";
import { zIndexDropdownable } from "../../design-tokens/build/js/designTokens";

export enum Direction {
  BottomLeft = "bottom-left",
  BottomRight = "bottom-right",
  BottomCenter = "bottom-center",
  TopLeft = "top-left",
  TopRight = "top-right",
  TopCenter = "top-center"
}

const DEFAULT_DIRECTION_PREFERENCES = [
  Direction.BottomLeft,
  Direction.BottomRight,
  Direction.TopRight,
  Direction.TopLeft
];

const Container = styled("div")`
  position: relative;
`;

export interface DropdownableProps {
  open: boolean;
  dropdown: React.ReactElement<any>;
  preferredDirections?: Direction[];
  onClose?: () => void;
  overlayRoot?: HTMLElement;
}

interface PositionCoord {
  top: number;
  left: number;
}

interface State {
  direction: Direction;
  position: PositionCoord;
}

interface Dimension {
  width: number;
  height: number;
}

const METHODS_TO_BIND = [
  "getDropdown",
  "setPosition",
  "setPositionFromCurrentProps",
  "calculateBestDirection",
  "positionForDirection",
  "childBounds",
  "windowDimensions",
  "dropdownDimensions"
];

class Dropdownable extends React.Component<DropdownableProps, State> {
  public static defaultProps: Partial<DropdownableProps> = {
    preferredDirections: DEFAULT_DIRECTION_PREFERENCES
  };

  private child = React.createRef<HTMLDivElement>();
  private dropdown = React.createRef<HTMLDivElement>();

  constructor(props) {
    super(props);

    this.state = {
      direction: Direction.BottomLeft,
      position: {
        top: 0,
        left: 0
      }
    };

    METHODS_TO_BIND.forEach(method => {
      this[method] = this[method].bind(this);
    });
  }

  componentDidMount() {
    resizeEventManager.add(this.setPositionFromCurrentProps);
    this.setPositionFromCurrentProps();
  }

  componentWillUnmount() {
    resizeEventManager.remove(this.setPositionFromCurrentProps);
  }

  UNSAFE_componentWillReceiveProps(nextProps: DropdownableProps) {
    if (nextProps.open) {
      this.setPosition(nextProps);
    }
  }

  render() {
    return (
      <Container>
        <div ref={this.child}>{this.props.children}</div>
        {this.getOverlay()}
      </Container>
    );
  }

  getOverlay() {
    const { position } = this.state;
    const { open, overlayRoot } = this.props;

    const overlayStyle = css({
      top: `${position.top}px`,
      left: `${position.left}px`,
      position: "absolute",
      transform: `scale(${open ? 1 : 0})`,
      opacity: open ? 1 : 0,
      "z-index": zIndexDropdownable
    });
    return (
      <Overlay overlayRoot={overlayRoot} className={overlayStyle}>
        {this.getDropdown()}
      </Overlay>
    );
  }

  getDropdown() {
    const { open, dropdown, onClose } = this.props;

    return (
      <div ref={this.dropdown}>
        <DropdownContents open={open} onClose={onClose}>
          {React.cloneElement(dropdown, {
            direction: this.state.direction
          })}
        </DropdownContents>
      </div>
    );
  }

  setPositionFromCurrentProps() {
    if (!this.props.open) {
      return;
    }
    this.setPosition(this.props);
  }

  setPosition(props: DropdownableProps) {
    const dropdownDimensions: Dimension = this.dropdownDimensions();
    const windowDimensions: Dimension = this.windowDimensions();
    const childBounds: ClientRect = this.childBounds();
    const direction: Direction = this.calculateBestDirection(
      props.preferredDirections,
      childBounds,
      dropdownDimensions,
      windowDimensions
    );

    this.setState({
      direction,
      position: this.positionForDirection(
        direction,
        childBounds,
        dropdownDimensions
      )
    });
  }

  calculateBestDirection(
    preferredDirections,
    childBounds,
    dropdownDimensions,
    windowDimensions
  ): Direction {
    // Determine if there is enough space in each direction to fit dropdown
    const triggerCenter = childBounds.width / 2 + childBounds.left;
    const possibleVertDirections = {
      top: dropdownDimensions.height <= childBounds.top,
      bottom:
        childBounds.bottom + dropdownDimensions.height <=
        windowDimensions.height
    };
    const possibleHorizDirections = {
      left:
        dropdownDimensions.width + childBounds.left <= windowDimensions.width,
      right: childBounds.right - dropdownDimensions.width >= 0,
      center:
        triggerCenter - dropdownDimensions.width / 2 > 0 &&
        triggerCenter + dropdownDimensions.width / 2 < windowDimensions.width
    };

    // Pick the first available preference
    const preferredDirection = preferredDirections.find(direction => {
      const topOrBottom = direction.split("-", 2)[0];
      const leftOrRight = direction.split("-", 2)[1];
      return (
        possibleVertDirections[topOrBottom] &&
        possibleHorizDirections[leftOrRight]
      );
    });

    // If nothing fits, fall back to the first preference
    return preferredDirection || preferredDirections[0];
  }

  positionForDirection(
    direction: Direction,
    childBounds,
    dropdownDimensions
  ): PositionCoord {
    const isTop: boolean =
      direction === Direction.TopLeft ||
      direction === Direction.TopRight ||
      direction === Direction.TopCenter;
    const isLeft: boolean =
      direction === Direction.TopLeft || direction === Direction.BottomLeft;
    const isCenter: boolean =
      direction === Direction.TopCenter || direction === Direction.BottomCenter;

    const getHorizAlignment = () => {
      if (isCenter) {
        return (
          childBounds.x +
          childBounds.width / 2 -
          dropdownDimensions.width / 2 +
          window.scrollX
        );
      } else if (isLeft) {
        return childBounds.left + window.scrollX;
      } else {
        return (
          childBounds.left -
          dropdownDimensions.width +
          childBounds.width +
          window.scrollX
        );
      }
    };

    return {
      top: isTop
        ? childBounds.top - dropdownDimensions.height + window.scrollY
        : childBounds.top + childBounds.height + window.scrollY,
      left: getHorizAlignment()
    };
  }

  childBounds(): ClientRect {
    const child = this.child.current;

    if (child) {
      return child.getBoundingClientRect();
    }

    // child.current should not be null due to React lifecycle but
    // is not guarateed by the type.
    return {
      width: 0,
      height: 0,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0
    };
  }

  windowDimensions(): Dimension {
    const body = document.body;
    return { width: body.clientWidth, height: body.clientHeight };
  }

  dropdownDimensions(): Dimension {
    const dropdown = this.dropdown.current!;

    if (dropdown) {
      return {
        width: dropdown.clientWidth,
        height: dropdown.clientHeight
      };
    }

    // dropdown.current should not be null due to React lifecycle but
    // is not guarateed by the type.
    return {
      width: 0,
      height: 0
    };
  }
}

export default Dropdownable;
