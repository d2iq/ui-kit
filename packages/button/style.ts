import { css } from "@emotion/css";
import {
  borderRadiusSmall,
  borderRadiusDefault,
  themeBrandPrimary,
  themeSuccess,
  themeError,
  themeTextColorInteractive,
  themeTextColorInteractiveInverted,
  themeTextColorPrimary,
  themeTextColorPrimaryInverted,
  themeBgDisabled,
  themeBgDisabledInverted,
  themeTextColorDisabled,
  themeBgNeutral,
  themeBgNeutralInverted,
  themeBrandPrimaryInverted,
  themeSuccessInverted,
  themeErrorInverted,
  themeTextColorDisabledInverted
} from "../design-tokens/build/js/designTokens";
import { darken, pickReadableTextColor } from "../shared/styles/color";
import { ButtonAppearances } from "./components/ButtonBase";
import { tintContent } from "../shared/styles/styleUtils";
import getCSSVarValue from "../utilities/getCSSVarValue";

export const buttonPadding = {
  vert: "10px",
  horiz: "18px"
};

const getHoverColor = color => darken(color, 1);
const getActiveColor = color => darken(color, 2);

export const filledButton = (
  baseColor: string,
  hoverColor: string,
  activeColor: string,
  isInverse?: boolean
) => {
  const contentColor = isInverse
    ? themeTextColorPrimaryInverted
    : pickReadableTextColor(
        baseColor,
        getCSSVarValue(themeTextColorPrimary),
        getCSSVarValue(themeTextColorPrimaryInverted)
      );

  return css`
    ${tintContent(contentColor)};
    background-color: ${baseColor};
    border-radius: ${borderRadiusSmall};
    padding: ${buttonPadding.vert} ${buttonPadding.horiz};

    &[href],
    &[href]:visited {
      ${tintContent(contentColor)};
    }

    &:hover {
      background-color: ${hoverColor};
    }

    &:active {
      background-color: ${activeColor};
    }
  `;
};

const mutedButton = css`
  ${tintContent(themeTextColorDisabled)};
  cursor: default;
  pointer-events: none;

  &:hover,
  &:focus,
  &:active {
    ${tintContent(themeTextColorDisabled)};
  }
`;

const inverseMutedButton = css`
  ${tintContent(themeTextColorDisabledInverted)};
  cursor: default;

  &:hover,
  &:focus,
  &:active {
    ${tintContent(themeTextColorDisabledInverted)};
  }
`;

const mutedFilledButton = css`
  background-color: ${themeBgDisabled};
  &:hover,
  &:focus,
  &:active {
    background-color: ${themeBgDisabled};
  }
`;

const inverseMutedFilledButton = css`
  background-color: ${themeBgDisabledInverted};
  &:hover,
  &:focus,
  &:active {
    background-color: ${themeBgDisabledInverted};
  }
`;

export const processingTextStyle = css`
  &:after {
    content: "...";
  }
`;

export const buttonBase = css`
  cursor: pointer;
  display: inline-block;
  outline: none;
  text-decoration: none;
`;

export const buttonContent = css`
  outline: none;
`;

export const fullWidthButton = css`
  text-align: center;
  width: 100%;
`;

export const focusStyles = (
  bgColor: string,
  borderColor: string = bgColor
) => css`
  outline: none;
  position: relative;

  &:focus {
    background-color: ${bgColor};

    &:after {
      border: 2px solid ${borderColor};
      border-radius: ${borderRadiusDefault};
      bottom: -3px;
      content: "";
      left: -3px;
      position: absolute;
      right: -3px;
      top: -3px;
    }
  }
`;

export const focusStyleByAppearance = (appearance, isInverse) => {
  switch (appearance) {
    case "primary":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeBrandPrimaryInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeBrandPrimary)));
    case "secondary":
      return css`
        ${isInverse
          ? focusStyles(
              "transparent",
              getHoverColor(getCSSVarValue(themeTextColorInteractiveInverted))
            )
          : focusStyles(
              "transparent",
              getHoverColor(getCSSVarValue(themeTextColorInteractive))
            )};
        &:focus {
          &:after {
            border-radius: 0;
            border-left-width: 0;
            border-right-width: 0;
            border-top-width: 0;
          }
        }
      `;
    case "standard":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeBgNeutralInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeBgNeutral)));
    case "success":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeSuccessInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeSuccess)));
    case "danger":
      return isInverse
        ? focusStyles(getHoverColor(getCSSVarValue(themeErrorInverted)))
        : focusStyles(getHoverColor(getCSSVarValue(themeError)));
    default:
      return "";
  }
};

export const button = appearance => {
  switch (appearance) {
    case "primary":
      return filledButton(
        getCSSVarValue(themeBrandPrimary),
        getHoverColor(getCSSVarValue(themeBrandPrimary)),
        getActiveColor(getCSSVarValue(themeBrandPrimary))
      );
    case "secondary":
      return css`
        ${tintContent(themeTextColorInteractive)};
        &:hover {
          ${tintContent(
            getHoverColor(getCSSVarValue(themeTextColorInteractive))
          )};
        }
        &:active {
          ${tintContent(
            getActiveColor(getCSSVarValue(themeTextColorInteractive))
          )};
        }
        &[href],
        &[href]:visited {
          ${tintContent(themeTextColorInteractive)};
        }
      `;
    case "standard":
      return filledButton(
        getCSSVarValue(themeBgNeutral),
        getHoverColor(getCSSVarValue(themeBgNeutral)),
        getActiveColor(getCSSVarValue(themeBgNeutral))
      );
    case "success":
      return filledButton(
        getCSSVarValue(themeSuccess),
        getHoverColor(getCSSVarValue(themeSuccess)),
        getActiveColor(getCSSVarValue(themeSuccess))
      );
    case "danger":
      return filledButton(
        getCSSVarValue(themeError),
        getHoverColor(getCSSVarValue(themeError)),
        getActiveColor(getCSSVarValue(themeError))
      );
    default:
      return "";
  }
};

export const buttonInverse = appearance => {
  switch (appearance) {
    case "primary":
      return filledButton(
        getCSSVarValue(themeBrandPrimaryInverted),
        getHoverColor(getCSSVarValue(themeBrandPrimaryInverted)),
        getActiveColor(getCSSVarValue(themeBrandPrimaryInverted))
      );
    case "secondary":
      return css`
        ${tintContent(themeTextColorInteractiveInverted)};
        &:hover {
          ${tintContent(
            getHoverColor(getCSSVarValue(themeTextColorInteractiveInverted))
          )};
        }
        &:active {
          ${tintContent(
            getActiveColor(getCSSVarValue(themeTextColorInteractiveInverted))
          )};
        }
        &[href],
        &[href]:visited {
          ${tintContent(themeTextColorInteractiveInverted)};
        }
      `;
    case "standard":
      return filledButton(
        getCSSVarValue(themeBgNeutralInverted),
        getHoverColor(getCSSVarValue(themeBgNeutralInverted)),
        getActiveColor(getCSSVarValue(themeBgNeutralInverted))
      );
    case "success":
      return filledButton(
        getCSSVarValue(themeSuccessInverted),
        getHoverColor(getCSSVarValue(themeSuccessInverted)),
        getActiveColor(getCSSVarValue(themeSuccessInverted))
      );
    case "danger":
      return filledButton(
        getCSSVarValue(themeErrorInverted),
        getHoverColor(getCSSVarValue(themeErrorInverted)),
        getActiveColor(getCSSVarValue(themeErrorInverted))
      );
    default:
      return "";
  }
};

export const getMutedButtonStyles = (appearance: ButtonAppearances) => {
  return appearance !== "secondary"
    ? css`
        ${mutedButton};
        ${mutedFilledButton};
      `
    : mutedButton;
};

export const getInverseMutedButtonStyles = (appearance: ButtonAppearances) => {
  return appearance !== "secondary"
    ? css`
        ${inverseMutedButton};
        ${inverseMutedFilledButton};
      `
    : inverseMutedButton;
};

export const pointerCursor = css`
  cursor: pointer;
`;

// Replicates default browser focus ring styles.
//
// The media query targets Webkit browsers, which can
// more accurately replicate the native focus ring style
export const keyboardFocus = css`
  > div:focus {
    outline: none;
  }

  &:focus > div {
    outline-color: Highlight;
    outline-width: thin;

    @media (-webkit-min-device-pixel-ratio: 0) {
      outline-color: -webkit-focus-ring-color;
      outline-style: auto;
      outline-width: unset;
    }
  }
`;
