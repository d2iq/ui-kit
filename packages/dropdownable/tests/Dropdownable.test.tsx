import React from "react";
import serializer from "@emotion/jest";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";

const mockEventManager = {
  add: jest.fn(),
  remove: jest.fn()
};

jest.doMock("../../utilities/resizeEventManager", () => {
  return mockEventManager;
});

import Dropdownable from "../components/Dropdownable";
import { PrimaryButton } from "../../button";

expect.addSnapshotSerializer(serializer);

describe("Dropdownable", () => {
  it("is visible after opening", () => {
    const component = mount(
      <Dropdownable isOpen={false} dropdown={<p>I'm a touchdown</p>}>
        <PrimaryButton>Click me</PrimaryButton>
      </Dropdownable>
    );

    expect(toJson(component)).toMatchSnapshot();
    component.setProps({ isOpen: true });
    expect(toJson(component)).toMatchSnapshot();
  });
});
