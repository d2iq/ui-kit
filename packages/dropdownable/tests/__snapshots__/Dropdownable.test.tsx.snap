// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdownable is visible after opening 1`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #FFFFFF;
  fill: #FFFFFF;
  background-color: #7D58FF;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  background-color: #704fe5;
}

.emotion-0:active {
  background-color: #6446cc;
}

<Dropdownable
  dropdown={
    <p>
      I'm a touchdown
    </p>
  }
  isOpen={false}
>
  <div>
    <PrimaryButton>
      <ButtonBase
        appearance="primary"
        data-cy="primaryButton"
      >
        <FocusStyleManager
          focusEnabledClass="css-xcotv6"
        >
          <button
            className="emotion-0"
            data-cy="primaryButton"
            onClick={[Function]}
            tabIndex={0}
            type="button"
          >
            <span>
              Click me
            </span>
          </button>
        </FocusStyleManager>
      </ButtonBase>
    </PrimaryButton>
  </div>
</Dropdownable>
`;

exports[`Dropdownable is visible after opening 2`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #FFFFFF;
  fill: #FFFFFF;
  background-color: #7D58FF;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  background-color: #704fe5;
}

.emotion-0:active {
  background-color: #6446cc;
}

<Dropdownable
  dropdown={
    <p>
      I'm a touchdown
    </p>
  }
  isOpen={true}
>
  <div>
    <PrimaryButton>
      <ButtonBase
        appearance="primary"
        data-cy="primaryButton"
      >
        <FocusStyleManager
          focusEnabledClass="css-xcotv6"
        >
          <button
            className="emotion-0"
            data-cy="primaryButton"
            onClick={[Function]}
            tabIndex={0}
            type="button"
          >
            <span>
              Click me
            </span>
          </button>
        </FocusStyleManager>
      </ButtonBase>
    </PrimaryButton>
  </div>
  <clickOutside(DropdownContents)
    isOpen={true}
  >
    <DropdownContents
      isOpen={true}
    >
      <ForwardRef
        style={
          Object {
            "left": "0",
            "position": "absolute",
            "top": "0",
            "zIndex": 2,
          }
        }
      >
        <Overlay
          innerRef={[Function]}
          overlayRoot={
            <body>
              <div>
                <div
                  style="z-index: 2; position: absolute; left: 0px; top: 0px;"
                >
                  <p>
                    I'm a touchdown
                  </p>
                </div>
              </div>
            </body>
          }
          style={
            Object {
              "left": "0",
              "position": "absolute",
              "top": "0",
              "zIndex": 2,
            }
          }
        >
          <Portal
            containerInfo={
              <div>
                <div
                  style="z-index: 2; position: absolute; left: 0px; top: 0px;"
                >
                  <p>
                    I'm a touchdown
                  </p>
                </div>
              </div>
            }
          >
            <div
              style={
                Object {
                  "left": "0",
                  "position": "absolute",
                  "top": "0",
                  "zIndex": 2,
                }
              }
            >
              <p>
                I'm a touchdown
              </p>
            </div>
          </Portal>
        </Overlay>
      </ForwardRef>
    </DropdownContents>
  </clickOutside(DropdownContents)>
</Dropdownable>
`;
